### DISCUSSION TITLE
Improve vitest doc

### DISCUSSION THREAD

**Discussion by VictoriqueMoe** [TypedProject/ts-express-decorators#2642] [Ideas]
I wrote a post here: https://x.com/VictoriqueM/status/1772221942868615380?s=20

Basically, when using unplugin-swc to bundle vitest, it made (on esm) debugging impossible, it seems that for some reason, no matter how hard i tried (supplying swc settings directly or inherit from tsconfig). sourcemaps where not emited correctly, my breakpoint and debugger was all over the place.

after much trial and error trying to find an ESbuild compatible bundler for decorator metadata, i found that using @rollup/plugin-typescript worked perfectly and now i can acctually debug code.

If anyone is having issues using vitest on ESM with the debugger. do the following:

```bash
npm install @rollup/plugin-typescript --save-dev
```

```ts
// vitest.config.ts
import { configDefaults, defineConfig } from "vitest/config";
import typescript from "@rollup/plugin-typescript";

export default defineConfig({
    test: {
        globals: true,
        root: "./",
        exclude: [...configDefaults.exclude, "node_modules/*"],
    },
    plugins: [typescript()],
    esbuild: false,
});
```
Maybe the docs should be updated to reflect this. just an idea

**Comment by nakedmcse** [comment]
This solves a long standing issue for us with being able to effectively test.

**Reply by Romakita** [reply]
You should not hesitate to open topics for discussion in these cases. Additionally, I only added vitest support a week or two ago :)

**Comment by Romakita** [comment]
@VictoriqueMoe Thanks a lot for the feedback.

It seems to be strange you have an issue with unplugin-swc while it's the recommended plugin by Nest.js here: https://docs.nestjs.com/recipes/swc#vitest

I just tested a fresh project a with ESM stack and it seems to works as expected:

![Capture d’écran 2024-03-25 à 13 46 59](https://github.com/tsedio/tsed/assets/1763311/0bc02c8e-b97b-4a26-8044-e70365e5f12d)

It could be interesting to investigate @VictoriqueMoe. Have you a small repository to reproduce that ?

**Reply by VictoriqueMoe** [reply]
small? no. but you can see the issue 
![image](https://github.com/tsedio/tsed/assets/27996712/b8063cd9-c799-4b6e-8434-f77f86c1a696)
the var section is weird, but the issue i am having is breakpoints are just on weird lines that make no sense..

Here is a video of using `unplugin-swc` and then i cahnge to use the typescript rollup and breakpoints work fine.

[video](https://www.dropbox.com/scl/fi/9ynn0ge4hrc6709u07zn4/Recording-2024-03-25-131714.mp4?rlkey=gvyqfqpka9szvhrfoqb5slgc0&raw=1)

If you want to see this yourself

clone this repo: 
https://github.com/waifuvault/WaifuVault
create a `.test.env` on root with content:
```
 PORT=localhost:8082
NODE_ENV=test
FILE_SIZE_UPLOAD_LIMIT_MB=512
BLOCKED_MIME_TYPES=application/x-dosexec,application/x-executable,application/x-hdf5,application/x-java-archive,application/vnd.rar
USE_CLOUDFLARE=false
MAX_URL_LENGTH=4096
```
and run `src\services\__test__\unit\FileCleaner.spec.ts`

**Reply by VictoriqueMoe** [reply]
if you try and debug `should process expired Files with success` and step into `fileCleaner.processFiles()` you will see your debugger is basically broken

**Reply by Romakita** [reply]
Ok here is my repo: https://github.com/Romakita/tsed-vitest-unplugin-swc

You can try to add breakpoint here:
![Capture d’écran 2024-03-25 à 15 16 06](https://github.com/tsedio/tsed/assets/1763311/30c87265-7aee-4caf-ae86-b357ef9af1a3)

tell me if you have the same issue ;). I'll take time to checkout your repository.

**Reply by VictoriqueMoe** [reply]
Ok, after looking at your seup, i found the issue. 

the first was that i needed `"@swc/helpers` ?(i guess for `externalHelpers` but maybe i don't really need that).

and the 2nd was that I had `minify` set to true AND 
```json
"minify": {
  "compress": {
	"unused": true
  },
  "mangle": {
	"topLevel": false
  }
},
```
in my `.swc` config.

so my config was:
```json
{
  "isModule": true,
  "minify": true,
  "jsc": {
    "minify": {
      "compress": {
        "unused": true
      },
      "mangle": {
        "topLevel": false
      }
    },
    "target": "esnext",
    "parser": {
      "syntax": "typescript",
      "tsx": false,
      "decorators": true,
      "dynamicImport": true,
      "importMeta": true
    },
    "transform": {
      "legacyDecorator": true,
      "decoratorMetadata": true
    }
  }
}
```
setting `minify` to `false` and removing `jsc.minify` fixed the problem.

so after the change, my `.swc` looks like
```json
{
  "isModule": true,
  "minify": false,
  "sourceMaps": true,
  "jsc": {
    "target": "esnext",
    "parser": {
      "syntax": "typescript",
      "tsx": false,
      "decorators": true,
      "dynamicImport": true,
      "importMeta": true
    },
    "transform": {
      "legacyDecorator": true,
      "decoratorMetadata": true
    }
  }
}
```
this is OK, but i really wanted miniy for my production build (i uise swc to build the JS for my project). maybe i can somhow use a different swc for testing and compile.

After making these config changes, debug works as expected

**Reply by VictoriqueMoe** [reply]
it's weird, cos you's expect the sourcemaps to work on minified code but setting even `sourceMaps` to true while having minify doesn't work. oh, also, i did not need the helper stuff, just the minify changes from `.swc`

**Reply by Romakita** [reply]
I agree the isn't ideal to have a wrong sourceMaps in production. But, minifying backend code for a nodejs context isn't really necessary. Maybe you will have a code more compressed and image size (container) more smaller, but the file is available in the container or local file system, the performance obtained in this case is very negligible.

It's possible to disable minify option of vite plugin like that (and give all necessary option for the plugin):

```ts
    swc.vite({
      // Explicitly set the module type to avoid inheriting this value from a `.swcrc` config file
      module: {type: "es6"},
      minify: false
    })
```

**Reply by VictoriqueMoe** [reply]
the filesize is smaller by a great margine. also adding `minify: false` to the swc vite plugin does not seem to override the global one, i thought it might, but actually, it doesn't seem to at all, adding it in does nothing. 

I have a choice now, keep the Typescript plugin that works with minify (as it effectivally users tsc to make the source file and not swc) and sacrafice some time to compile for tests or remove minify from swc and sacrafice like 5mb from the image. 

I think i will stick with Typescript plugin instead of swc.

**Reply by Romakita** [reply]
Hum ok. I’ll take a time to read the plugin code. 
the given example by nest.js have comment that said:
> // Explicitly set the module type to avoid inheriting this value from a `.swcrc` config file

I hope we can define an entire swc config for the vitest plugin. If not, we can open PR to fix that and hope that the change will merged quickly ^^.

**Reply by VictoriqueMoe** [reply]
I saw that too, but no matter what i put in the plugin config, it doesn't seem to effect. the plugin seems to inherit from tsconfig, but not `swc` itself. no sure why, it's weird. I can prove this by basically setting crap in the swc plugin and everything still works, in fact, you don't even need `type: "es6"` 

Yea, It would be nice to have both work i agree! 

I tried this:
```ts
import swc from "unplugin-swc";
import { defineConfig } from "vitest/config";

export default defineConfig({
    test: {
        globals: true,
        root: "./",
    },
    plugins: [
        swc.vite({
            minify: false,
            sourceMaps: true,
            jsc: {
                minify: {},
            },
            module: {
                type: "es6",
            },
        }),
    ],
});
```
but it didn't do anything

**Reply by Romakita** [reply]
Good news. This configuration seems to works as expected:

```ts
import swc from "unplugin-swc";
import { defineConfig } from "vitest/config";

export default defineConfig({
    test: {
        globals: true,
        root: "./",
    },
    plugins: [
        swc.vite({
            sourceMaps: true,
            jsc: {
                minify: {
                    sourceMap: true,
                    mangle: false,
                    compress: false,
                    keep_classnames: true,
                    keep_fnames: true,
                },
            },
            module: {
                type: "es6",
            },
        }),
    ],
});

```

**Reply by Romakita** [reply]
I let you check ;)

**Reply by Romakita** [reply]
If it doesn't works, I suggest your to rename the **.swcrc** to **.swcrc.prod** and change the build command:

```ts
{
    "build": "npm run barrels && node ./buildScripts/cleanDist.mjs && swc src -d dist --config-file .swcrc.prod --copy-files && node ./buildScripts/moveFiles.mjs",
}
```
See you :)

**Reply by VictoriqueMoe** [reply]
> Good news. This configuration seems to works as expected:
> 
> ```ts
> import swc from "unplugin-swc";
> import { defineConfig } from "vitest/config";
> 
> export default defineConfig({
>     test: {
>         globals: true,
>         root: "./",
>     },
>     plugins: [
>         swc.vite({
>             sourceMaps: true,
>             jsc: {
>                 minify: {
>                     sourceMap: true,
>                     mangle: false,
>                     compress: false,
>                     keep_classnames: true,
>                     keep_fnames: true,
>                 },
>             },
>             module: {
>                 type: "es6",
>             },
>         }),
>     ],
> });
> ```

This did not work, it made it slightly better, but the debug was still messed up for `fileCleaner.processFiles`.

i changed my config to 
```ts
import swc from "unplugin-swc";
import { configDefaults, defineConfig } from "vitest/config";

export default defineConfig({
    test: {
        globals: true,
        root: "./",
        exclude: [...configDefaults.exclude, "node_modules/*", "src/__test__"],
    },
    plugins: [
        swc.vite({
            module: {
                type: "es6",
            },
        }),
    ],
});
```
and did that you suggested by renaming my build command and now it works fine. so it was my `.swc` that for some reason, the plugin was not overriding correctly, so renaming that works now.
